### ISSUE TITLE
FIDO2 device not recognised

### ISSUE THREAD

**Issue by heylix** [mjec/fido2-hmac-secret#23]
**Describe the bug**
`khefin enumerate` tells me `Device at /dev/hidraw5 is not a FIDO2 authenticator` even though it is.

**Expected behavior**
Khefin should recognise the FIDO2 stick and tell me that it's not supported.

**Environment:**
 - Fedora 32
 - Version: 0.4.3

**Authenticator make and model:**
- [Hypersecu HyperFIDO](https://hypersecu.com/tmp/products/hyperfido) Mini
- Supports FIDO U2F, FIDO2 and OATH HOTP
- lsusb says: `2ccf:0880 HS HyperFIDO Token`

**Comment by mjec** [comment]
Hi @heylix, thanks for filing this report!

I think there are two distinct issues here:

 1. `enumerate` shouldn't exit prematurely if the attached device isn't FIDO2 compatible, we should just flag that device as not supported.
 2. your specific device is being flagged as not FIDO2 compatible, although it's certified to be so.

I have just pushed a fix for (1) to master, but will need more information to figure out (2).

We use [libfido2](https://github.com/Yubico/libfido2) to determine whether the requested device is a FIDO2 device, and that error indicates that `fido_dev_is_fido2(device)` returns `false`.

To see why that's happening, please run `FIDO_DEBUG=1 khefin enumerate` and send me the output you get on STDERR. Note that may include sensitive information, so rather than attaching it to this bug report you can email it to me at khefin@mjec.net.

**Comment by heylix** [comment]
Pulled from master again, compiled (now khefin reports itself as version 0.5.0) and got this:

```
$ FIDO_DEBUG=1 ./khefin enumerate
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
fido_tx: d=0x1f784a0, cmd=0x86, buf=0x1f784a0, count=8
  43 1d 40 ac 72 d8 37 c5 
fido_rx: initiation frame at 0x7ffedb558e80, len 64
  ff ff ff ff 86 00 11 43 1d 40 ac 72 d8 37 c5 00 
  00 00 01 02 01 00 01 03 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
fido_dev_get_cbor_info_tx: dev=0x1f784a0
fido_tx: d=0x1f784a0, cmd=0x90, buf=0x7ffedb558f37, count=1
  04 
fido_dev_get_cbor_info_rx: dev=0x1f784a0, ci=0x1f78ec0, ms=-1
fido_rx: initiation frame at 0x7ffedb558ca0, len 64
  00 00 00 01 bf 00 01 01 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
fido_rx: cid (0x1000000, 0x1000000), cmd (0xbf, 0x90)
fido_dev_get_cbor_info_rx: fido_rx
khefin: Unable to get info from device: FIDO_ERR_RX (0xfffffffe)
```
(I don't think this includes any sensitive information, does it?)

**Comment by heylix** [comment]
Seems that there's a [possible fix for this](https://github.com/Yubico/libfido2/commit/ba946883087e3121bcf57098da63196f8b039a4d) in libfido2 v.1.4.0, while I'm running 1.3.1 from the F32 repositories.

<del>I'll try with libfido2 v1.4.0 and report back!</del>

Slightly different output with libfido2 v1.4.0:

```
$ FIDO_DEBUG=1 ./khefin enumerate
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
fido_tx: d=0x515ae0, cmd=0x06, buf=0x515ae0, count=8
0000: ed 8d e4 a0 05 c6 cc 14
fido_rx: d=0x515ae0, cmd=0x06, buf=0x515ae8, count=17, ms=-1
rx_preamble: initiation frame at 0x7ffc6fcb5030
0000: ff ff ff ff 86 00 11 ed 8d e4 a0 05 c6 cc 14 00
0016: 00 00 07 02 01 00 01 03 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=17
fido_rx: buf=0x515ae8, len=17
0000: ed 8d e4 a0 05 c6 cc 14 00 00 00 07 02 01 00 01
0016: 03
fido_dev_get_cbor_info_tx: dev=0x515ae0
fido_tx: d=0x515ae0, cmd=0x10, buf=0x7ffc6fcb50d7, count=1
0000: 04
fido_dev_get_cbor_info_rx: dev=0x515ae0, ci=0x514af0, ms=-1
fido_rx: d=0x515ae0, cmd=0x10, buf=0x7ffc6fcb4c10, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc6fcb4b80
0000: 00 00 00 07 bf 00 01 01 00 00 00 00 00 00 00 00
0016: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx_preamble: cid (0x7000000, 0x7000000), cmd (0xbf, 0x10)
rx: rx_preamble
fido_dev_get_cbor_info_rx: fido_rx
khefin: Unable to get info from device: FIDO_ERR_RX (0xfffffffe)
```

**Comment by mjec** [comment]
This is really surprising. libfido2 should come with a tool called `fido2-token` -- can you try running that as `fido2-token -I /dev/hidraw5` and see what you get back? 

Based on your debug transcript it seems like your device doesn't support the mandatory CTAPHID_CBOR command, which would explain why it's erroring out like this. If `fido2-token` works though, this is something funny happening in khefin which I'll have to debug.

I appreciate you providing all this extra info, it's super helpful!

<details><summary>How I figured out CTAPHID_CBOR is the issue</summary>

This is your authenticator's response to the [CTAPHID_INIT](https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#usb-hid-init) command:

```
0000: ed 8d e4 a0 05 c6 cc 14 00 00 00 07 02 01 00 01
0016: 03
```

The 02 means the device supports CTAP2, but after that things go a little off the rails. The trailing 03 is an illegal value; it should be a bitmask with 0x02 set to zero and 0x04 set to one. That the 4 bit is set to zero means the authenticator does not implement CTAPHID_CBOR, even though that's a mandatory command for CTAP2.

The next command issued by libfido2 is a CTAPHID_CBOR command regardless (0x10), and it's not surprising that fails given the device does not support it. The response _almost_ indicates as much: 0xbf for error and code 01 for unsupported command. In fact 0xbf is for bluetooth devices; error responses from a USB device should be 0x3f, which explains why this is handled as a generic error rather than as device unsupported.
</details>

**Comment by heylix** [comment]
>can you try running that as `fido2-token -I /dev/hidraw5` and see what you get back?

sure!

```
$ fido2-token -dI /dev/hidraw5

fido_tx: d=0x55beb66052a0, cmd=0x06, buf=0x55beb66052a0, count=8
0000: 13 ae 42 92 92 a1 4b 20
fido_rx: d=0x55beb66052a0, cmd=0x06, buf=0x55beb66052a8, count=17, ms=-1
rx_preamble: initiation frame at 0x7ffef68876c0
0000: ff ff ff ff 86 00 11 13 ae 42 92 92 a1 4b 20 00
0016: 00 00 03 02 01 00 01 03 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=17
fido_rx: buf=0x55beb66052a8, len=17
0000: 13 ae 42 92 92 a1 4b 20 00 00 00 03 02 01 00 01
0016: 03
proto: 0x02
major: 0x01
minor: 0x00
build: 0x01
caps: 0x03 (wink, nocbor, msg)
```

**Comment by heylix** [comment]
Apparently the device isn't certified to be FIDO2 compliant. Why does it still report to support CTAP2?

I looked up the Hypersecu products on the [FIDO2 certified products list](https://fidoalliance.org/certification/fido-certified-products/):

```
Company Name: Hypersecu Information Systems, Inc.
Implementation Name: Hyper FIDO® Mini
Specification: U2F
Version: 1.0
Type: Authenticator
Authenticator Level: Functional Only
```

The only FIDO2 certified product by that company seems to be the Hyperfido Pro:

```
Company Name: Hypersecu Information Systems, Inc.
Implementation Name: Hyper FIDO® Pro
Specification: FIDO2
Version: 2.0
Type: Authenticator
Authenticator Level: Level 1
```

Can this still somehow made to work with khefin or is all lost? It's really weird to me that some sites claim that the device supports FIDO2 but clearly there's no CTAPHID_CBOR capability as displayed in the libfido2 output above…

EDIT: I have successfully created and verified a credential with `fido2-cred` via the U2F method, but couldn't get libfido2 to create a credential with the FIDO2 method. All the data seems to point to the fact that the Hypersecu HyperFIDO Mini doesn't support FIDO2.
